src/mito2/src/compaction/memory_manager.rs

```diff
@@ -0,0 +1,285 @@
+    semaphore: Semaphore,
+    limit_bytes: u64,
+    used_bytes: AtomicU64,
+    notify: Notify,
+}
+
+impl CompactionMemoryManager {
+    /// Creates a new memory manager with the given limit in bytes.
+    /// `limit_bytes = 0` disables the limit.
+    pub fn new(limit_bytes: u64) -> Self {
+        if limit_bytes == 0 {
+            COMPACTION_MEMORY_LIMIT.set(0);
+            return Self { inner: None };
+        }
+
+        let permits = bytes_to_permits(limit_bytes) as usize;
+        COMPACTION_MEMORY_LIMIT.set(limit_bytes as i64);
+
+        Self {
+            inner: Some(Arc::new(Inner {
+                semaphore: Semaphore::new(permits),
+                limit_bytes,
+                used_bytes: AtomicU64::new(0),
+                notify: Notify::new(),
+            })),
+        }
+    }
+
+    /// Returns the configured limit in bytes (0 if unlimited).
+    pub fn limit_bytes(&self) -> u64 {
+        self.inner
+            .as_ref()
+            .map(|inner| inner.limit_bytes)
+            .unwrap_or(0)
+    }
+
+    /// Returns currently used bytes.
+    pub fn used_bytes(&self) -> u64 {
+        self.inner
+            .as_ref()
+            .map(|inner| inner.used_bytes.load(Ordering::Relaxed))
+            .unwrap_or(0)
+    }
+
+    /// Returns available bytes.
+    pub fn available_bytes(&self) -> u64 {
+        self.limit_bytes().saturating_sub(self.used_bytes())
+    }
+
+    /// Tries to acquire memory. Returns Some(guard) on success, None if insufficient.
+    pub fn try_acquire(&self, bytes: u64) -> Option<CompactionMemoryGuard> {
+        match &self.inner {
+            None => Some(CompactionMemoryGuard::unlimited()),
+            Some(inner) => {
+                let permits = bytes_to_permits(bytes);
+                if permits == 0 {
+                    return Some(CompactionMemoryGuard::unlimited());
+                }
+
+                match inner.semaphore.try_acquire_many(permits) {
+                    Ok(permit) => {
+                        permit.forget();
+                        inner.on_acquire(permits);
+                        Some(CompactionMemoryGuard::limited(inner.clone(), permits))
+                    }
+                    Err(TryAcquireError::NoPermits) | Err(TryAcquireError::Closed) => {
+                        COMPACTION_MEMORY_REJECTED
+                            .with_label_values(&["try_acquire"])
+                            .inc();
+                        None
+                    }
+                }
+            }
+        }
+    }
+
+    /// Waits until the specified amount of memory becomes available.
+    ///
+    /// Note: This method does not reserve the memory. After this method returns,
+    /// you should call `try_acquire` to attempt to reserve the memory. However,
+    /// there is no guarantee the memory will still be available at that point,
+    /// as another task may acquire it concurrently.
+    pub async fn wait_for_available(&self, bytes: u64) {
```

Comment:
Why not use https://docs.rs/tokio/latest/tokio/sync/struct.Semaphore.html#method.acquire

Reply:
The purpose here is only to detect, but it will not consume the permits. This is because the design only consumes permits in one place, which is before executing `submit_compaction_task`.